### PR TITLE
Profile envelopes sent directly from profiler

### PR DESCRIPTION
**Pull request by stefanosiano**
## :scroll: Description
The `AndroidTransactionProfiler` now directly sends profile envelopes. This means we don't have to pass the `ProfileTraceData` anymore anyway. 
So we could drop the old timed out profiles handling, as they are directly sent when they time out

## :bulb: Motivation and Context
This change allows profiles to be sent independently from transactions.
It also fixes and simplify timed out profiles handling. Previously we had to wait for a transaction to finish to attach the timed out profile to it and send the envelope. Now the envelope is sent as soon as the profile times out.

## :green_heart: How did you test it?
I added a ui test to check the profile is sent on timeout without finishing the transaction
I added few unit tests to check the `captureProfile` method of the hub is called by the profiler itself.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by stefanosiano** [comment]
Do I need a changelog entry? The `ITransactionProfiler` class is internal, and so should be all the changes in this pr

**Comment by markushi** [comment]
> Do I need a changelog entry? The `ITransactionProfiler` class is internal, and so should be all the changes in this pr

Good question! @romtsn how do we usually deal with internal changes?

**Comment by markushi** [comment]
Looks good to me! 👍

**Comment by romtsn** [comment]
I guess the changelog is not only for the users but also for us, right? So I'd probably add an entry there. Also, this sounds like a behavior change (profiles are sent without transactions now), which makes it reasonable to add to the changelog again.

File: sentry/src/main/java/io/sentry/SentryTracer.java
Comment (romtsn) [inline sentry/src/main/java/io/sentry/SentryTracer.java:370]: there was no test for this previously? Sounds like it should fail

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java
Comment (romtsn) [inline sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java:284]: How about we move this into the constructor? Then we should be also able to test it properly without overriding `Sentry.setCurrentHub`

Perhaps, we could even make `AndroidTransactionProfiler` an integration, but I'm not sure if it was ever discussed before @marandaneto

**Review by adinauer — APPROVED** [review]
LGTM, agree with roman on the changelog. I too think it should be listed.

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java:284]: Maybe create an envelope out of the `profilingTraceData` and call `captureEnvelope` instead?
Your call here @romtsn 
I don't recall any discussion related to `AndroidTransactionProfiler` as an integration.

File: sentry/src/main/java/io/sentry/SentryTracer.java
Comment (stefanosiano) [reply sentry/src/main/java/io/sentry/SentryTracer.java:370]: previously it was captured with the transaction, and the only required parameter was the transaction (the others could be null)
then I switched to send the profile directly using `hub.captureProfile` and added the null check
now i moved the `captureProfile` call into the Profiler directly
the test check if finishing a transaction calls the `profiler.onTransactionFinish` method, and the profiler tests check if the `ISentryClient.captureProfile` is called

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java
Comment (stefanosiano) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java:284]: Out of curiosity, why is it preferred to create an Envelope and call `captureEnvelope` instead of calling `captureProfile`/`captureTransaction`/`captureWhatever`?
Is it for consistency or are there other reasons? @marandaneto 
I can quickly create the envelope and drop the `captureProfile` methods, if it's better

Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java:284]: Because this expands the public API and is not needed.

Comment (stefanosiano) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java:284]: @adinauer  Actually, `HubAdapter` is public, and it has a `captureProfile` method

Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java:284]: `HubAdapter` is just to make Sentry static class testable.

Comment (stefanosiano) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java:284]: That's fine, but `HubAdapter` isn't marked as internal, so technically there's nothing preventing users to actually use it.
Just wanted to make sure we are aware of this

Comment (stefanosiano) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java]: @marandaneto i made the change. Would you take another look before merging?

**Review by romtsn — APPROVED** [review]
@stefanosiano I'm planning to cut a release today, would you like to merge it so it gets in?

**Pull request merged**
